### ISSUE TITLE
Can't connect to device via adb

### ISSUE THREAD

**Issue by tincturaginseng** [rumpeltux/viomi-rooting#31]
OS: Arch Linux x86_64
Device: Viomi Vacuum Cleaner Mop P (STYTJ02YM)
Firmware: **3.5.8_0017**
ADB and USB cable works fine (tested on my android phone).

1. I disassembled the device, disconnected the battery, connected the USB cable.
2. Run script (and press Enter) and wait
3. Plug USB cable to USB port on PC ant watch in **journalctl -xf**
```
Aug 16 11:45:05 localhost kernel: usb 2-1.5: new high-speed USB device number 4 using ehci-pci
Aug 16 11:45:05 localhost kernel: usb 2-1.5: New USB device found, idVendor=1f3a, idProduct=efe8, bcdDevice=ff.ff
Aug 16 11:45:05 localhost kernel: usb 2-1.5: New USB device strings: Mfr=0, Product=0, SerialNumber=0
Aug 16 11:45:08 localhost kernel: usb 2-1.5: USB disconnect, device number 4
```
After that I hear "Turning on" from device, waiting 2 seconds and unplug USB cable from PC. Script still waiting to connect to device.
I'll try this more than 10 times - no result. Device disconnected after 3 seconds after connecting and script don't go to next step.

Maybe this firmware is non-rootable?

**Comment by mkozinenko** [comment]
Are you sure about FW version? 3.5.3_0017 maybe? Guys [here](https://github.com/rumpeltux/viomi-rooting/issues/29) reporting that it works. Though I have the same situation with yours with 3.5.3_0017 firmware.

**Comment by tincturaginseng** [comment]
> Are you sure about FW version? 3.5.3_0017 maybe?
I'm sure, FW **3.5.8_0017**
![Screenshot_20210817-234644](https://user-images.githubusercontent.com/3119849/129798520-a81dc425-a316-4ca1-8f69-a8b526e1392c.jpg)
I disable mtp-probe from udev rules, stop and disable VMware services (network, USB, etc). No results.
I assume that the script does not have time to connect to the device during its boot.

**Comment by Manoz** [comment]
Same issue here with STYTJ02YM. Here's all the photos I took when disassembling my robot : https://imgur.com/gallery/3X8UgnE

**Comment by tincturaginseng** [comment]
> Hi I have the same issue. It seems there is no ADB exposed via the usb port under the battery. The viomi SE has another port under the rubber plug on the bottom, but my viomi V8 does not. I am now thinking about opening the robot to look for another port. Did you maybe already do that? Then I could keep my warranty seal...

A vacuum cleaner is defined as a device when a cable is connected to an existing USB port (I see this via lsusb). But ADB does not appear on this port or the script does not have time to connect to it

**Comment by razserv2010** [comment]
I have the same problem, anyone found a solution?

**Comment by benoegen** [comment]
Here: https://github.com/rumpeltux/python-miio/issues/1#issuecomment-915647117

Basically hold home and press power ten times after having plugged in usb. It will then expose adb and I just ran the script and it worked.

**Comment by Dropaq** [comment]
so. this method is for newer revisions manufactured in 2021
according to this comment https://github.com/rumpeltux/viomi-rooting/issues/32#issuecomment-911237365 robots from 2020 still had the ADB exposed on normal boot

though, apparently the script is still compatible as we have an ADB now, please be aware, that it is doing a little too much, like killing the robot manager and then restoring it. If it fails in the process - robot might not work and needs to be manually fixed (https://github.com/rumpeltux/viomi-rooting/issues/32)

@benoegen which errors you had during execution?

maybe we can make a separate version of the rooting-tool to simplify the process

**Comment by benoegen** [comment]
Problem for me was that I am not too familiar with linux commands in the first place, so I did not know if I have to paste the whole command block or every command by itself. So I was already unsure if I created the adb_shell correctly. Removing the password by "sed" command did not work, but that was the only "real error".

Basically this was the reason to use the script, which worked fine for me.

**Comment by razserv2010** [comment]
> Here: [rumpeltux/python-miio#1 (comment)](https://github.com/rumpeltux/python-miio/issues/1#issuecomment-915647117)
> 
> Basically hold home and press power ten times after having plugged in usb. It will then expose adb and I just ran the script and it worked.

I did what you said but the vacuum still sounds "turn on"

**Comment by benoegen** [comment]
Did you read the linked comment? After plugging in, I waited a short moment, then started clicking power but took it rather slow, I think it took two attempts. What output gives `dmesg `for you?

>     Keep USB connected to the robot, but not to the PC
>     Long press the power key for at least 10 seconds to power off the device
>     Press the "Home" key and do not release it.
>     Connect the USB to the PC
>     Click power key for about 10 times
>     Release both keys
>     Robot is going to boot into FEL mode, and start the ADB
> 
> > dmesg
> should show you an android device popping up.
> The voice will confirm the connection as well.

**Comment by razserv2010** [comment]
> Did you read the linked comment? After plugging in, I waited a short moment, then started clicking power but took it rather slow, I think it took two attempts. What output gives `dmesg `for you?
> 
> > ```
> > Keep USB connected to the robot, but not to the PC
> > Long press the power key for at least 10 seconds to power off the device
> > Press the "Home" key and do not release it.
> > Connect the USB to the PC
> > Click power key for about 10 times
> > Release both keys
> > Robot is going to boot into FEL mode, and start the ADB
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > > dmesg
> > > should show you an android device popping up.
> > > The voice will confirm the connection as well.

[ 5921.521759] usb 1-1.1: new high-speed USB device number 35 using xhci_hcd
[ 5921.652002] usb 1-1.1: New USB device found, idVendor=1f3a, idProduct=efe8, bcdDevice=ff.ff
[ 5921.652024] usb 1-1.1: New USB device strings: Mfr=0, Product=0, SerialNumber=0
[ 5924.495506] usb 1-1.1: USB disconnect, device number 35
[ 5924.691675] Voltage normalised (0x00000000)

**Comment by razserv2010** [comment]
I did not understand "Release both keys"
In the last 10 click on power on
Should I also release the home button?

**Comment by benoegen** [comment]
You keep the home key pressed the whole time. You plug in USB, click power 10x, the home key is pressed the whole time, after clicking the power key the last time, you release both, power and home

**Comment by tincturaginseng** [comment]
> You press home and keep it pressed. You plug in USB, click power 10x, the home key is pressed the whole time, after clicking the power key the last time, you release both, power and home.
> 
> After that it takes some seconds for the device to show up as android under `dmesg `

For fw 3.5.8_0017 this method not work. My steps:

1. Turn off device
2. Disconnect then battery
3. Connect micro-USB cable to robot
4. Hold down **Home** key and connect USB cable to PC
5. Press **Power** key for 10 times and release both - **Home** and **Power** keys
6. In _lsusb_ output I see _Allwinner Technology sunxi SoC OTG connector in FEL/flashing mode_ device for 3 seconds and it disconnect. Output for _adb devices_ still empty
7. After that I hear "turning on" message and robot try to normal boot

Video for process - https://user-images.githubusercontent.com/3119849/132990528-172776da-22b4-4a8d-b58a-11b9d70eeb57.mp4

**Comment by colordeficiency** [comment]
> > You press home and keep it pressed. You plug in USB, click power 10x, the home key is pressed the whole time, after clicking the power key the last time, you release both, power and home.
> > After that it takes some seconds for the device to show up as android under `dmesg `
> 
> For fw 3.5.8_0017 this method not work. My steps:
> 
>     1. Turn off device
> 
>     2. Disconnect then battery
> 
>     3. Connect micro-USB cable to robot
> 
>     4. Hold down **Home** key and connect USB cable to PC
> 
>     5. Press **Power** key for 10 times and release both - **Home** and **Power** keys
> 
>     6. In _lsusb_ output I see _Allwinner Technology sunxi SoC OTG connector in FEL/flashing mode_ device for 3 seconds and it disconnect. Output for _adb devices_ still empty
> 
>     7. After that I hear "turning on" message and robot try to normal boot
> 
> 
> Video for process - https://user-images.githubusercontent.com/3119849/132990528-172776da-22b4-4a8d-b58a-11b9d70eeb57.mp4

I am having the same issues.

`> lsusb
> Bus 001 Device 044: ID 1f3a:efe8 Allwinner Technology sunxi SoC OTG connector in FEL/flashing mode

> dmesg
[Sep13 02:57] usb 1-1: new high-speed USB device number 44 using xhci_hcd
[  +0.152220] usb 1-1: New USB device found, idVendor=1f3a, idProduct=efe8, bcdDevice=ff.ff
[  +0.000013] usb 1-1: New USB device strings: Mfr=0, Product=0, SerialNumber=0
[  +2.785393] usb 1-1: USB disconnect, device number 44`

Any idea what should the lights be when adb is enabled?

**Comment by prometheus247** [comment]
Same problem here. 

> > > You press home and keep it pressed. You plug in USB, click power 10x, the home key is pressed the whole time, after clicking the power key the last time, you release both, power and home.
> > > After that it takes some seconds for the device to show up as android under `dmesg `
> > 
> > 
> > For fw 3.5.8_0017 this method not work. My steps:
> > ```
> > 1. Turn off device
> > 
> > 2. Disconnect then battery
> > 
> > 3. Connect micro-USB cable to robot
> > 
> > 4. Hold down **Home** key and connect USB cable to PC
> > 
> > 5. Press **Power** key for 10 times and release both - **Home** and **Power** keys
> > 
> > 6. In _lsusb_ output I see _Allwinner Technology sunxi SoC OTG connector in FEL/flashing mode_ device for 3 seconds and it disconnect. Output for _adb devices_ still empty
> > 
> > 7. After that I hear "turning on" message and robot try to normal boot
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Video for process - https://user-images.githubusercontent.com/3119849/132990528-172776da-22b4-4a8d-b58a-11b9d70eeb57.mp4
> 
> I am having the same issues.
> 
> ```
> > lsusb
> > Bus 001 Device 044: ID 1f3a:efe8 Allwinner Technology sunxi SoC OTG connector in FEL/flashing mode
> 
> > dmesg
> [Sep13 02:57] usb 1-1: new high-speed USB device number 44 using xhci_hcd
> [  +0.152220] usb 1-1: New USB device found, idVendor=1f3a, idProduct=efe8, bcdDevice=ff.ff
> [  +0.000013] usb 1-1: New USB device strings: Mfr=0, Product=0, SerialNumber=0
> [  +2.785393] usb 1-1: USB disconnect, device number 44
> ```
> 
> Any idea what should the lights be when adb is enabled?

**Comment by patrickelectric** [comment]
any updates ?

**Comment by gdumgdum** [comment]
Having the exact same problem here.

After fully disassembling and looking at the motherboard, it looks like the extra USB port that other robots have under the rubber is a duplicate of the port next to the battery (it's not a second USB port, it's the exact same USB port, pin to pin). On other robots this mirror port is connected to the 5-pin socket labelled J32.

In my case, the only noticeable difference is when I power off the robot, keep pressed the Home button, plug the USB cable, hold Power for 4 seconds until the leds switch from slow-blinking-white to steady red+white, then release both buttons.
Only after these steps, the robots says "turning on", then 3 seconds later "device connected, setup is now completed".
However, no ADB device shows up, only the "Allwinner Technology sunxi SoC OTG connector" device for 3 seconds at the very beginning of the process.

Still investigating.

**Comment by patrickelectric** [comment]
just to update here, I got it working in the end, I don't remember how. But I got it holding the buttons, watching dmesg and running the script. The `ADB_ARG` was necessary for me.

**Comment by gdumgdum** [comment]
> just to update here, I got it working in the end, I don't remember how. But I got it holding the buttons, watching dmesg and running the script. The `ADB_ARG` was necessary for me.

`ADB_ARG` ?

**Comment by gdumgdum** [comment]
Ok, I succeeded but it wasn't easy. Here is what worked for me on my Mijia STYTJ02YM viomi.vacuum.v8:

- a RaspberryPi, powered by a strong 5V power supply (Nintendo Switch power supply in my case)
- a good USB/micro USB cable compatible with data transfer (believe it or not, quality is important, I used different cables and some where not able to provide enough power to the robot motherboard, which rebooted randomly)
- I used a USB 2.0 port of the Raspberry, not a USB 3.0 port (not sure this is important)
- Robot battery unplugged (this is important), usb cable plugged in the motherboard but not on the Raspberry yet.
- Procedure:
  - boot the Raspberry, open 2 consoles
  - type `journalctl -kf` in one of the consoles, it will display useful messages about USB device detections
  - in the second console, download (or git clone) the `https://github.com/rumpeltux/viomi-rooting.git` and launch `NEW_V8=1 ./viomi-root.sh`. Validate until it displays the procedure to boot the robot in FEL mode.
  - Press and keep pressed the robot Home button
  - Plug the usb cable to a USB 2.0 port of the Raspberry (keep an eye on the journal logs)
  - still pressing Home, press Power until leds light up white, then change. Either they turn orange and you're good, either they turn red+white, then you can unplug the USB cable and retry (I can't remember if I kept Power pressed or if I quickly pressed it 10/15 times, I made so many attempts...)
  - if the journal logs shows a device with something like `idVendor=1f3a, idProduct=efe8`, it failed. Don't loose time and restart.
  - if the journal logs shows a device with something like `idVendor=18d1, idProduct=d002` and `USB Developer`, it worked! Release both buttons and proceed with the script.

Note: if the motherboard never speaks, your USB cable is not good enough.
Note 2: the viomi-root.sh scripts needs to be modified in order to change the dropbear configuration. In function `install_dropbear()`, right before the `EOF` line, add this:
`    PubkeyAcceptedAlgorithms +ssh-rsa`
`    HostkeyAlgorithms +ssh-rsa`
Note3: dismantling the robot helps a lot to see if the leds are orange or red+white (you can't really tell the difference unless you directly see the leds)

**Comment by aurismat** [comment]
Apologies for potentially necroing this issue, but I'd like to note down my own successes and failures in getting root and installing Valetudo on my own Xiaomi Vacuum Cleaner Mop P (same model of STYTJ02YM), FW `3.5.3_0017`(Mi Home app wouldn't let me update beyond that) - Home Assistant in Viomi's integration was marking it as `viomi.vacuum.v8` - maybe it'll help someone or give ideas in making experimental v8 procedures less 'experimental'.

My problem was, that no matter how many times I'd attempt different button combinations, plugging procedures, it would not go forward.

Eventually I did more of the manual approach described here - [`itooktheredpill`'s blogpost](https://itooktheredpill.irgendwo.org/2020/rooting-xiaomi-vacuum-robot/) where I specifically called:
```
adb shell /etc/init.d/robotManager disable
```
in a loop while having the motherboard with no battery and just the MicroUSB cable plugged in - just to wait for that window before robotApp disables ADB.

While I understand removing the rc.d link and doing this effectively achieves the same thing, disabling it via init.d produced an useable ADB shell *for me* from which I was able to get dropbear set up and afterwards, follow the procedures on [Hypfer/valetudo-crl200s-root](https://github.com/Hypfer/valetudo-crl200s-root/blob/master/README.md) to install the dustbuilder's firmware + Valetudo binaries.

Hope this helps!

EDIT: in case someone needs the full snippet to loop for ADB:
```sh
while true; do
    adb shell /etc/init.d/robotManager disable 2>/dev/null && break
done
```
